javascript-algorithms-and-data-structures/learn-basic-string-and-array-methods-by-building-a-music-player-app/script.js

```diff
@@ -0,0 +1,326 @@
+    if(document.activeElement.classList.contains('playlist-song')) {
+      const nextSongDivs = document.querySelectorAll(".playlist-song");
+      nextSongDivs[currentIndex].focus();
+    }
+  }
+
+  function previousSong() {
+    const prevIndex = (currentIndex - 1 + songs.length) % songs.length;
+    playSong(songs[prevIndex]);
+    currentIndex = prevIndex;
+
+  //   keeping keyboard indicator in sync with array keyls/selected indicator.
+    if(document.activeElement.classList.contains('playlist-song')) {
+        const previousSongDivs = document.querySelectorAll(".playlist-song");
+        previousSongDivs[currentIndex].focus();
+      }
+  }
+
+  function renderSongDisplay(currentIndex) {
+    document.getElementById("player-song-title").innerText = songs[currentIndex].title;
+    document.getElementById("player-song-artist").innerText = songs[currentIndex].artist;
+    document.getElementById("player-album-art").innerHTML = `<img src="${songs[currentIndex].cover}" alt="song cover art" />`;
+  }
+
+
+  // playlist display
+  function clearSongBgs() {
+      document.querySelectorAll(".playlist-song").forEach(song => {
+          song.removeAttribute('aria-current');
+      });
+  }
+
+  function updatePlayingSongBackground(newIndex) {
+      const updateSongDivs = document.querySelectorAll(".playlist-song");
+      updateSongDivs[newIndex].setAttribute('aria-current', true);
+  }
+
+  function playSelectedSong(song, songBtn) {
+    playSong(song);
+    clearSongBgs();
+    songBtn.setAttribute('aria-current', true);
+  }
+
+
+  function shuffle() {
+    // first time the spread operator is being taught. Make sure to introduce with example in the steps
+    let songsCopy = [...songs];
+    // using array sort method and Math.random method
+    songsCopy.sort(() => 0.5 - Math.random());
+    renderSongs(songsCopy);
+    // reset eventlisteners to stay in sync with songs playing
+    setSongEventListener();
+  }
+
+  function deleteSong(song){
+    let resetBtn = document.createElement('button');
+    let resetText = document.createTextNode('Reset Playlist');
+    resetBtn.appendChild(resetText);
+    resetBtn.ariaLabel = 'reset playlist';
+    resetBtn.classList.add('player-reset-btn');
+    resetBtn.onclick = function () {
+      app()
+    };
+    // first time splice is introduced. make sure to provide explanation and example in the steps.
+    songs.splice(song,1);
+    renderSongs(songs);
+    setDeleteEventListener();
+
+    if(songs.length < 2) {
+      songListDiv.append(resetBtn);
+    }
+  }
+
+  // Event listeners
+  document.querySelector(".play").addEventListener("click", togglePlay);
+  document.querySelector(".pause").addEventListener("click", pauseSong);
+  document.querySelector(".previous").addEventListener("click", previousSong);
+  document.querySelector(".next").addEventListener("click", nextSong);
+  document.querySelector(".shuffle").addEventListener("click", shuffle);
```

Comment:
I am having hard time getting the shuffle button to work here.

Reply:
While testing locally, if I start a song and then click on the shuffle button then the indicator goes back to the first song but it doesn't look like the list is shuffled

https://github.com/freeCodeCamp/CurriculumExpansion/assets/67210629/a176d111-4809-4f63-95be-6cafd2c14d7d